### ISSUE TITLE
TypeNameRule should validate associatedtype

### ISSUE THREAD

**Issue by marcelofabri** [realm/SwiftLint#956]
Something similar to https://github.com/realm/SwiftLint/pull/955 would be needed.

```swift
protocol P {
    associatedtype X
}
```

```bash
$ sourcekitten structure --file file.swift                          
{
  "key.substructure" : [
    {
      "key.kind" : "source.lang.swift.decl.protocol",
      "key.offset" : 0,
      "key.nameoffset" : 9,
      "key.namelength" : 1,
      "key.bodyoffset" : 12,
      "key.bodylength" : 22,
      "key.accessibility" : "source.lang.swift.accessibility.internal",
      "key.runtime_name" : "_TtP8__main__1P_",
      "key.name" : "P",
      "key.length" : 35
    }
  ],
  "key.offset" : 0,
  "key.diagnostic_stage" : "source.diagnostic.stage.swift.parse",
  "key.length" : 36
}

$ sourcekitten syntax --file file.swift
[
  {
    "offset" : 0,
    "length" : 8,
    "type" : "source.lang.swift.syntaxtype.keyword"
  },
  {
    "offset" : 9,
    "length" : 1,
    "type" : "source.lang.swift.syntaxtype.identifier"
  },
  {
    "offset" : 17,
    "length" : 14,
    "type" : "source.lang.swift.syntaxtype.keyword"
  },
  {
    "offset" : 32,
    "length" : 1,
    "type" : "source.lang.swift.syntaxtype.identifier"
  }
]
```

**Comment by marcelofabri** [comment]
Fixed in #955
